### ISSUE TITLE
wait for process with pid=pid to exit

### ISSUE THREAD

**Issue by ORESoftware** [nodejs/help#1132]
This applies to Node.js versions > 6

I am looking for something like this in C#
https://msdn.microsoft.com/en-us/library/fb4aw7b8(v=vs.110).aspx

in Node.js, that might look like:

```js
process.waitForExit(pid, function(err, code){
  // err might be thrown if pid is not live
  // code is exit code of process with pid
});
```

**Comment by gireeshpunathil** [comment]
the [child_process's async methods](https://nodejs.org/dist/latest-v9.x/docs/api/child_process.html#child_process_child_process_spawn_command_args_options) that handle child processes inherently have these capabilities right?

```js
const { spawn } = require('child_process')
const ls = spawn('ls', ['-lrt'])

ls.on('error', (err) => {
  console.log(`process threw error:  ${err}`)
})

ls.on('close', (code) => {
  console.log(`process exited with code ${code}`)
})
```

**Comment by ORESoftware** [comment]
sorry, I wasn't clear - the process that I want to wait for is _not_ a child process of my node process - all I have is the pid.

I found an answer here:
https://stackoverflow.com/a/41613532/1223975

so I have this:

```js
const cmd = isDarwin ?
  `lsof -p ${grandParentPid} +r 1 &>/dev/null` :
  `tail --pid=${grandParentPid} -f /dev/null`;
```


actually fairly hard to find a solution. if you use the `wait` command it won't work because the process has to be a child process of the current one.

**Comment by the1mills** [comment]
Here are some relevant threads on StackExchange if youre interested:

https://unix.stackexchange.com/q/427115/113238

https://unix.stackexchange.com/q/427255/113238

**Comment by the1mills** [comment]
This might work too:
https://stackoverflow.com/a/39231660/1223975

**Comment by ORESoftware** [comment]
@gireeshpunathil I think there might be a way to do this with dtrace in Node.js, what do you think?

**Comment by ORESoftware** [comment]
dtrace should be able to "attach" to another process x...my dtrace watcher just needs to exit, or get some signal, when the other process x dies.

**Comment by ORESoftware** [comment]
https://developer.apple.com/library/content/technotes/tn2050/_index.html

**Comment by gireeshpunathil** [comment]
@ORESoftware - as I said, there is no standard / portable way of watching a foreign process. So if you are looking for a feature request, you may back it with a use case, and I guess will be considered for implementation based on the strength of the use case.

**Comment by ORESoftware** [comment]
yeah it would be a cool feature for node.js core, here is what is needed for both macos/linux:

https://github.com/ORESoftware/quicklock/tree/master/src

the linux file works for linux, the darwin file works for macos

**Comment by gireeshpunathil** [comment]
@ORESoftware - with your code I got 
`ptrace(PTRACE_SEIZE, 32335, NULL, NULL) = -1 EPERM (Operation not permitted)`
this is because ptrace on arbitrary process are inhibited by some OS settings for example [ubuntu's kernel hardening](https://wiki.ubuntu.com/SecurityTeam/Roadmap/KernelHardening#ptrace)

**Comment by gireeshpunathil** [comment]
@ORESoftware - so my bottom line is that while it is technically possible to wait on an alien process termination, the implementation effort would be high due to lack of system-wide compatible APIs, and needs to be weighed in based on the `strength` and `popularity` of the use case. Will you please share them, for the team's consideration? thanks!

Or else if you are happy with the explanations and happy to close this let me know.

**Comment by ORESoftware** [comment]
sure, the implementations I have should work well without polling etc...where do I share?

**Comment by gireeshpunathil** [comment]
here itself please. I have added `enhancement` label.

**Comment by gireeshpunathil** [comment]
@ORESoftware raised feature request in https://github.com/nodejs/node/issues/19207 . Closing this in favor of that where the discussion is active.

**Comment by ORESoftware** [comment]
👍
